core/src/broadcast_stage.rs

```diff
@@ -385,25 +385,41 @@ pub fn broadcast_shreds(
         return Ok(());
     }
     let mut shred_select = Measure::start("shred_select");
+    let stakes: Vec<_> = peers_and_stakes.iter().map(|(stake, _)| *stake).collect();
     let packets: Vec<_> = shreds
         .iter()
         .map(|shred| {
-            let broadcast_index = weighted_best(&peers_and_stakes, shred.seed());
-
-            (&shred.payload, &peers[broadcast_index].tvu)
+            let index: Vec<_> = weighted_shuffle(&stakes, shred.seed())
+                .into_iter()
+                .map(|i| peers_and_stakes[i].1)
+                .collect();
+            // Broadcast to the first node.
+            // Forward to the rest of the first layer.
+            let (neighbors, _) = compute_retransmit_peers(DATA_PLANE_FANOUT, 0, &index);
+            let anchor = &peers[neighbors[0]];
+            let forwards = neighbors
+                .into_iter()
+                .skip(1)
+                .map(|i| &peers[i].tvu_forwards)
+                .filter(|addr| ContactInfo::is_valid_address(addr));
+            let addrs: Vec<_> = std::iter::once(&anchor.tvu).chain(forwards).collect();
```

Comment:
hm. so the leader sends to the whole first layer?

Reply:
yes, and I think this is consistent with the intended design in:
https://docs.solana.com/cluster/turbine-block-propagation#layer-and-neighborhood-structure
> The current leader makes its initial broadcasts to at most DATA_PLANE_FANOUT nodes.

Reply:
Yes, but it does not send every shred to every node in the first layer. The first design is that it should do a round-robin of the nodes in the first layer with each shred, then those nodes transfer within themselves. Otherwise the leader has 200x (fanout times) the upload bandwidth requirement to send all the block data.

Reply:
> Yes, but it does not send every shred to every node in the first layer.

Yes, the current code does not do that. And I have mentioned in PR description this is a change I am making, which to me sounds more consistent with the intended design in:
https://docs.solana.com/cluster/turbine-block-propagation#layer-and-neighborhood-structure

Currently the leader broadcast the shred to exactly one node and that node can be any node, including an unstaked node (and the packet may be lost or the node may go down, or turbine peers computations may be inconsistent between the leader and the node which receives the packet). So my understanding is that this is very fragile and does not seem to have enough redundancy. If that single packet is lost only the leader will have the shred in the cluster.

> then those nodes transfer within themselves

The documentation suggests that that is the intended design. But the implementation is not doing that. Only the first node in each neighborhood is forwarding the shred to other nodes in that neighborhood. Other nodes in the neighborhood do not send packets to each other.

> Otherwise the leader has 200x (fanout times) the upload bandwidth requirement to send all the block data.

Yes, that is added cost of this PR. But already that single node which obtains the shred from the leader does that 200x fanout:
https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/retransmit_stage.rs#L387-L388

I am not happy with the 200x fanout cost for the leader either. Perhaps the `DATA_PLANE_FANOUT` could be smaller. Or for now the leader could forward to a sample of nodes in the first layer rather than all.

Reply:
No it's not the intended design. Erasure coding makes up for the fact that a single packet could get lost to a down node.

So if there are 200 shreds in a slot and fanout=200. The leader broadcasts a single shred to each node in layer 0, then each node in layer 0 broadcasts to the other nodes in layer 0. The leader would send 200 shreds total and each of the nodes in the first layer also sends 199 shreds (and maybe 1 to lower layers). So everyone has sent about the same amount and has optimal upload bandwidth usage. If the leader sends to all 200, then it will have sent many more shreds than everyone else (200 *200).

Reply:
To me, that documentation is vague and says: "The current leader makes its initial broadcasts to at most DATA_PLANE_FANOUT nodes." Which does not say exactly how the leader splits the broadcast between those 1st layer nodes. Just that the initial broadcasts should be to that set. Where does it say the leader broadcasts all data to all nodes?

This one is more specific:
https://medium.com/solana-labs/turbine-solanas-block-propagation-protocol-solves-the-scalability-trilemma-2ddba46a51db

"Our solution to this problem, Turbine, borrows heavily from BitTorrent — although a few major technical details differentiate the two. Turbine is optimized for streaming, and transmits data using UDP only, and implements a random path per packet through the network as leaders (block producers) stream their data. The leader breaks the block into packets up to 64KB in size. For a 128MB block, the leader produces 2,000 64KB packets, and transmits each packet to a different validator."

Reply:
This is what the current implementation is doing:
* The leader sends the shred to the first node in the first layer: https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/broadcast_stage.rs#L390-L394
* Each node always broadcasts to all its children (200 fanout): https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/retransmit_stage.rs#L386-L391
* The first node in each neighborhood on each layer also broadcasts to all its neighbor in its layer (so an additional 199 fanout): https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/retransmit_stage.rs#L387
* No other two nodes in the same layer or neighborhood send packets to each other.

So, effectively:
* The leader has fanout of: 1
* First node of each neighborhood in each layer has fanout of: 200 + 199 (children + neighbors)
* All other nodes have fanout of: 200 (children)

Reply:
I see. It's only 200 children if there are 200 neighborhoods in the next lower level right?

Reply:
yes

Reply:
Does the broadcast go to only the 1st layer? I thought it weights all tvu_peers? Is it filtered somewhere?

Reply:
Not sure I understood the question.
Currently, the leader sends the shred only to the first node in the first layer: https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/broadcast_stage.rs#L390-L394

From there it is re-transmitted to the rest of cluster.

Reply:
It sends to 1 random node in `peers_and_stakes` but doesn't that come from `get_broadcast_peers` here: https://github.com/solana-labs/solana/blob/33dc63ccbdd685e9b3abe35640fd6a563dac81e9/core/src/broadcast_stage.rs#L366-L369

That looks at `cluster_info.tvu_peers()` which is all nodes in the whole network with a valid TVU port right? So, it will not be the subset of nodes which are just in layer 1.

Reply:
Right, but `shred.seed()` is fixing how nodes are ordered in the turbine retransmit tree, and `wighted_best` here:
https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/broadcast_stage.rs#L390-L394
is effectively returning the very first node in the first layer.
It is equivalent of `shuffle_peers_and_index` in retransmit:
https://github.com/solana-labs/solana/blob/33dc63ccb/core/src/retransmit_stage.rs#L341
and then taking the very first node.

Reply:
Got it. Thanks.

Reply:
Can we have a decision here?

As I described above: https://github.com/solana-labs/solana/pull/14910#discussion_r567443834
current code is passing the onus of broadcasting from leader to the first node in the first layer (which is an extra hop and a single point of failure with no redundancy). See https://github.com/solana-labs/solana/issues/6594 for example:
> If the shred was originally sent to the node from slot leader (and it got dropped), likely **the whole cluster is missing it**.

I do not see how that is beneficial enough to do so. All nodes have a potential fanout of at least 200 whereas the leader is only 1.

That being said, the part that is broadcasting from leader to other nodes in the first layer, can be left out of this pr. The primary fix here is excluding the leader from peers computation in retransmit-stage to make it consistent with broadcast stage. You can use [turbine-tree-dump-branch](https://github.com/behzadnouri/solana/commit/turbine-tree-dump) which samples the turbine tree at broadcast and retransmit stage and shows that for the same value of the seed the two stages do **not** order the nodes in the same way. This is probably one of the reasons https://github.com/solana-labs/solana/pull/7774 got stuck.

Reply:
I don't think this strategy of broadcasting to everyone in layer 1 is feasible in the real network with fanout of 200. It works fine in a small network with nodes on the same 10gbps switch as tested, but 200x more upload bandwidth for the leader will be huge and way less efficient. We can scale the duplication with higher erasure rate. With erasure coding not any one data packet is a single point of failure, because there are multiple coding shreds sent that could recover the shred lost from that pathway.

> That being said, the part that is broadcasting from leader to other nodes in the first layer, can be left out of this pr

Ok, yes that sounds good.

Reply:
Reverted the part broadcasting from leader to first layer